pkg/config/setup/config.go

```diff
@@ -1765,38 +1767,104 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
-				// kubernetes_node_annotations_as_tags, ...). This is an issue when a setting use a
-				// map[string] as a value. The key can contain a '.' making it impossible to set or get.
-				//
-				// The secrets resolver doesn't have the notion of what is a known setting and what
-				// isn't. Returning false tells the secretResolver that we refuse the notification and
-				// we want to be notified for the parent key.
-				//
-				// See secrets.ResolveCallback documentation for more information.
-				//
-				if !config.IsKnown(settingName) {
-					return false
-				}
-
-				log.Debugf("replacing handle for setting '%s' with secret value", settingName)
-				config.Set(settingName, value, pkgconfigmodel.SourceAgentRuntime)
-				return true
-			})
-		if err != nil {
+		secretResolver.Subscribe(func(handle string, settingPath []string, oldValue, newValue any) {
+			if err := configAssignAtPath(config, settingPath, newValue); err != nil {
+				log.Errorf("could not assign to config: %s", err)
+			}
+		})
+		if _, err = secretResolver.Resolve(yamlConf, origin); err != nil {
 			return fmt.Errorf("unable to decrypt secret from datadog.yaml: %v", err)
 		}
 	}
 	return nil
 }
 
+func configAssignAtPath(config pkgconfigmodel.Config, settingPath []string, newValue any) error {
+	settingName := strings.Join(settingPath, ".")
+	if config.IsKnown(settingName) {
+		config.Set(settingName, newValue, pkgconfigmodel.SourceAgentRuntime)
+		return nil
+	}
+
+	// Trying to assign to an unknown config field can happen when trying to set a
+	// value inside of a compound object (a slice or a map) which allows arbitrary key
+	// values. Some settings where this happens include `additional_endpoints`, or
+	// `kubernetes_node_annotations_as_tags`, etc. Since these arbitrary keys can
+	// contain a '.' character, we are unable to use the standard `config.Set` method.
+	// Instead, we remove trailing elements from the end of the path until we find a known
+	// config field, retrieve the compound object at that point, and then use the trailing
+	// elements to figure out how to modify that particular object, before setting it back
+	// on the config.
+	//
+	// Example: configAssignAtPath({'process_config', 'additional_endpoints', 'http://url.com', '0'}, 'password')
+	//
+	// This is split into:
+	//   ['process_config', 'additional_endpoints']  // a known config field
+	// and:
+	//   ['http://url.com', '0']                     // trailing elements
+	//
+	// This function will effectively do:
+	//
+	// var original map[string][]string = config.Get('process_config.additional_endpoints')
+	// var slice []string               = original['http://url.com']
+	// slice[0] = 'password'
+	// config.Set('process_config.additional_endpoints', original)
+
+	trailingElements := []string{}
+	// copy the path and hold onto the original, useful for error messages
+	path := make([]string, len(settingPath))
+	copy(path, settingPath)
```

Comment:
Could use `slices.Clone` here (`path := slices.Clone(settingPath)`)

Reply:
Nice, I had searched for this but missed it in the `slices` package.